streams/src/main/java/org/apache/kafka/streams/kstream/internals/suppress/InMemoryTimeOrderedKeyValueBuffer.java

```diff
@@ -0,0 +1,130 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.kstream.internals.suppress;
+
+import org.apache.kafka.streams.KeyValue;
+
+import java.util.ConcurrentModificationException;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Objects;
+import java.util.TreeMap;
+
+class InMemoryTimeOrderedKeyValueBuffer<K, V> implements TimeOrderedKeyValueBuffer<K, V> {
+    private long memBufferSize = 0;
+    private final Map<K, TimeKey<K>> index = new HashMap<>();
+    private final TreeMap<TimeKey<K>, SizedValue<V>> sortedMap = new TreeMap<>();
+
+    private static class SizedValue<V> {
+        private final V v;
+        private final long size;
+
+        private SizedValue(final V v, final long size) {
+            this.v = v;
+            this.size = size;
+        }
+
+        public V value() {
+            return v;
+        }
+
+        public long size() {
+            return size;
+        }
+
+        @Override
+        public String toString() {
+            return "SizedValue{v=" + v + ", size=" + size + '}';
+        }
+    }
+
+    private class InnerIterator implements Iterator<KeyValue<TimeKey<K>, V>> {
+        private final Iterator<Map.Entry<TimeKey<K>, SizedValue<V>>> delegate = sortedMap.entrySet().iterator();
+        private Map.Entry<TimeKey<K>, SizedValue<V>> last;
+
+        @Override
+        public boolean hasNext() {
+            return delegate.hasNext();
+        }
+
+        @Override
+        public KeyValue<TimeKey<K>, V> next() {
+            final Map.Entry<TimeKey<K>, SizedValue<V>> next = delegate.next();
+            this.last = next;
+            return new KeyValue<>(next.getKey(), next.getValue().value());
+        }
+
+        @Override
+        public void remove() {
+            delegate.remove();
+            index.remove(last.getKey().key());
+            memBufferSize = memBufferSize - last.getValue().size();
+        }
+    }
+
+    @Override
+    public Iterator<KeyValue<TimeKey<K>, V>> iterator() {
+        return new InnerIterator();
+    }
+
+    @Override
+    public void put(final long time,
+                    final K key,
+                    final V value,
+                    final long recordSize) {
+        final TimeKey<K> previousKey = index.get(key);
+        // non-resetting semantics:
+        // if there was a previous version of the same record, insert the new record in the same place in the priority queue
+        final TimeKey<K> timeKey = previousKey != null ? previousKey : new TimeKey<>(time, key);
+
+        final TimeKey<K> replaced = index.put(key, timeKey);
+
+        if (!Objects.equals(previousKey, replaced)) {
```

Comment:
Do we need this check?

Reply:
I tend to think, that we don't need this guard because a bug that gives multi-threaded access seems to be very unlikely. But it's a personal opinion... My concern again is because this is the hot code path. But I am also ok to keep the check if somebody insists.

Reply:
I also think it's unlikely to be useful. I'll remove it.